[5.0] [Feature] support largest-triangle-three-buckets algo

SnailSword opened the pull request:
<!-- Please fill in the following information to help us review your PR more efficiently. -->

## Brief Information

This pull request is in the type of:

- [ ] bug fixing
- [x] new feature
- [ ] others



### What does this PR do?

<!-- USE ONCE SENTENCE TO DESCRIBE WHAT THIS PR DOES. -->

Support largest-triangle-three-buckets sampling algorithm.

### Fixed issues

<!--
- #xxxx: ...
-->

#9403 

## Usage

### Are there any API changes?

- [x] The API has been changed.

<!-- LIST THE API CHANGES HERE -->

sampling: 'lttb'

### Related test cases or examples to use the new APIs

```
{
  type: 'line',
  sampling: 'lttb',
  data: data
}
```

See **test/line-large.html** line 101.

## Others

### Merging options

- [ ] Please squash the commits into a single one when merge.

### Other information

In master branch it works, see #13312.
But in next branch, because of #12997 , samplers doesn't run in some cases. 
And when it run, there is a point highlight bug, I think it's better to fix those bugs in other pr.

pissang commented on src/data/List.ts:
```diff
@@ -1692,6 +1692,109 @@ class List<
+    /**
+     * Large data down sampling using largest-triangle-three-buckets
+     * copied from https://github.com/pingec/downsample-lttb with some modifications
+     * @param {string} baseDimension
```
Please don't use the code directly. It will have licence issue.

pissang commented:
Also, do you mind add a test to compare the result and performance difference between lttb sampling enabled and disabled?

The example used in https://github.com/apache/incubator-echarts/pull/13314#issuecomment-695758284 will be a good case.

SnailSword commented:
> Also, do you mind adding a test case to compare the result and performance difference between lttb sampling enabled and disabled?
> 
> The example used in [#13314 (comment)](https://github.com/apache/incubator-echarts/pull/13314#issuecomment-695758284) will be a good case.
> 

![demo](https://user-images.githubusercontent.com/19756301/93916339-f4ca0700-fd3b-11ea-8f38-adb820bd78b6.gif)


@pissang Thank you for your advise. I added comparison of results in `test/sample-compare.html`. But I don't know how to compare the performance, time spent on this part(the pic below) is almost the same in each downsampling method, and also same as that without downsampling. I don't think it is the correct way to compare the performance.

![image](https://user-images.githubusercontent.com/19756301/93914843-b0d60280-fd39-11ea-9ca5-59853ac9667d.png)



> And there are some code style like inconsistent indentation needs to be fixed.

Fixed.
